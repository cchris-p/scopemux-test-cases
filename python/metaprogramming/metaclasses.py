#!/usr/bin/env python3
"""
Demonstration of Python Metaprogramming

This example covers:
- Metaclasses
- Class decorators
- Dynamic attribute access (__getattr__, __setattr__)
- Descriptors
- Dynamic class creation (type)
- Attribute access customization
- Runtime class modification
"""

from typing import Any, Dict, List, Callable, Type, Optional, TypeVar, cast
import functools
import inspect
import time


def log_calls(cls):
    """A class decorator that logs method calls"""

    # Store the original __getattribute__
    original_getattribute = cls.__getattribute__

    # Define a new __getattribute__ that wraps method calls with logging
    def __getattribute__(self, name):
        attr = original_getattribute(self, name)
        if callable(attr) and not name.startswith("__"):

            @functools.wraps(attr)
            def wrapper(*args, **kwargs):
                print(f"Calling {cls.__name__}.{name}")
                result = attr(*args, **kwargs)
                print(f"Finished {cls.__name__}.{name}, returned {result}")
                return result

            return wrapper
        return attr

    # Replace the original __getattribute__ with our version
    cls.__getattribute__ = __getattribute__
    return cls


def singleton(cls):
    """A class decorator that implements the Singleton pattern"""

    instances = {}

    @functools.wraps(cls)
    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]

    return get_instance


class ValidateAttributes(type):
    """A metaclass that enforces type validation for attributes"""

    def __new__(mcs, name, bases, attrs):
        # Create _validators dictionary to store validation functions
        attrs["_validators"] = {}

        # Look for validate_<attr> methods and register them
        for key, value in list(attrs.items()):
            if key.startswith("validate_"):
                attr_name = key[9:]  # Remove 'validate_' prefix
                attrs["_validators"][attr_name] = value

        return super().__new__(mcs, name, bases, attrs)

    def __init__(cls, name, bases, attrs):
        super().__init__(name, bases, attrs)

        # Store original __setattr__
        original_setattr = cls.__setattr__

        # Define a new __setattr__ that validates attributes
        def __setattr__(self, name, value):
            if name in self._validators:
                # Call the validator method
                validator = getattr(self, f"validate_{name}")
                value = validator(value)

            # Call the original __setattr__
            original_setattr(self, name, value)

        # Replace __setattr__ with our version
        cls.__setattr__ = __setattr__


class LazyProperty:
    """A descriptor that computes the value only when first accessed"""

    def __init__(self, func):
        self.func = func
        self.__doc__ = func.__doc__
        self.__name__ = func.__name__

    def __get__(self, instance, owner):
        if instance is None:
            return self

        # Compute the value
        value = self.func(instance)

        # Cache the value on the instance
        setattr(instance, self.__name__, value)

        return value


class TimingDescriptor:
    """A descriptor that measures the execution time of methods"""

    def __init__(self, func):
        self.func = func
        functools.update_wrapper(self, func)

    def __get__(self, instance, owner):
        if instance is None:
            return self

        @functools.wraps(self.func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            result = self.func(instance, *args, **kwargs)
            end_time = time.time()
            print(f"{self.func.__name__} took {end_time - start_time:.6f} seconds")
            return result

        return wrapper


def timing_method(func):
    """A method decorator that measures execution time"""

    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        start_time = time.time()
        result = func(self, *args, **kwargs)
        end_time = time.time()
        print(f"{func.__name__} took {end_time - start_time:.6f} seconds")
        return result

    return wrapper


class AttributeProxy:
    """A descriptor that delegates attribute access to another object"""

    def __init__(self, target_attr):
        self.target_attr = target_attr

    def __get__(self, instance, owner):
        if instance is None:
            return self

        # Get the target object
        target = getattr(instance, self.target_attr)
        return target


# Example of a class with metaclass
class Person(metaclass=ValidateAttributes):
    """A person with validated attributes"""

    def __init__(self, name: str, age: int, email: str):
        self.name = name
        self.age = age
        self.email = email

    def validate_name(self, value):
        """Validate that name is a non-empty string"""
        if not isinstance(value, str) or not value:
            raise ValueError("Name must be a non-empty string")
        return value

    def validate_age(self, value):
        """Validate that age is a positive integer"""
        if not isinstance(value, int) or value <= 0:
            raise ValueError("Age must be a positive integer")
        return value

    def validate_email(self, value):
        """Validate that email contains @"""
        if not isinstance(value, str) or "@" not in value:
            raise ValueError("Email must be a string containing @")
        return value

    @LazyProperty
    def formatted_info(self):
        """Return formatted information about the person"""
        print("Computing formatted_info (this happens only once)")
        return f"{self.name} ({self.age}) - {self.email}"

    @timing_method
    def do_work(self, duration: float = 0.1):
        """Simulate doing some work"""
        time.sleep(duration)
        return f"{self.name} completed work"


# Example of a class with a class decorator
@log_calls
class Calculator:
    """A simple calculator with logging"""

    def add(self, a, b):
        return a + b

    def subtract(self, a, b):
        return a - b

    def multiply(self, a, b):
        return a * b

    def divide(self, a, b):
        if b == 0:
            raise ValueError("Division by zero")
        return a / b


# Example of a singleton class
@singleton
class AppConfig:
    """A singleton configuration class"""

    def __init__(self):
        self.settings = {}
        print("AppConfig instance created")

    def get_setting(self, key, default=None):
        return self.settings.get(key, default)

    def set_setting(self, key, value):
        self.settings[key] = value


class DynamicAttributes:
    """A class that handles missing attributes dynamically"""

    def __init__(self):
        self._attributes = {}

    def __getattr__(self, name):
        """Called when an attribute is not found"""
        if name in self._attributes:
            return self._attributes[name]

        # Create a dynamic method if asked for get_<name>
        if name.startswith("get_"):
            attr_name = name[4:]  # Remove 'get_' prefix
            return lambda: self._attributes.get(attr_name)

        raise AttributeError(f"{self.__class__.__name__} has no attribute '{name}'")

    def __setattr__(self, name, value):
        """Called when an attribute is set"""
        if name == "_attributes":
            # Allow setting the _attributes dictionary directly
            super().__setattr__(name, value)
        else:
            # Store other attributes in the _attributes dictionary
            self._attributes[name] = value


def demonstrate_metaclass():
    print("\n=== Metaclass Example ===")

    try:
        # Create a person with valid attributes
        person = Person("Alice", 30, "alice@example.com")
        print(f"Created person: {person.name}, {person.age}, {person.email}")

        # Access the lazy property
        print(f"Formatted info: {person.formatted_info}")
        # Access it again (should be cached)
        print(f"Formatted info again: {person.formatted_info}")

        # Try with invalid attributes
        try:
            person.age = -5
            print("This should not be printed")
        except ValueError as e:
            print(f"Validation error: {e}")

        # Call the timed method
        result = person.do_work()
        print(f"Work result: {result}")

    except Exception as e:
        print(f"Error: {e}")


def demonstrate_class_decorator():
    print("\n=== Class Decorator Example ===")

    # Create a calculator instance
    calc = Calculator()

    # Call methods (should be logged)
    result1 = calc.add(5, 3)
    print(f"5 + 3 = {result1}")

    result2 = calc.multiply(4, 6)
    print(f"4 * 6 = {result2}")


def demonstrate_singleton():
    print("\n=== Singleton Example ===")

    # Create multiple instances (should be the same object)
    config1 = AppConfig()
    config1.set_setting("theme", "dark")

    config2 = AppConfig()
    theme = config2.get_setting("theme")
    print(f"Theme from config2: {theme}")

    # Prove they're the same object
    print(f"config1 is config2: {config1 is config2}")


def demonstrate_dynamic_attributes():
    print("\n=== Dynamic Attributes Example ===")

    # Create an instance
    obj = DynamicAttributes()

    # Set some attributes
    obj.name = "Dynamic Object"
    obj.value = 42

    # Get attributes
    print(f"Name: {obj.name}")
    print(f"Value: {obj.value}")

    # Use a dynamic getter method
    get_name = obj.get_name
    print(f"get_name(): {get_name()}")

    # Try to access a non-existent attribute
    try:
        print(obj.nonexistent)
    except AttributeError as e:
        print(f"Error: {e}")


def demonstrate_dynamic_class_creation():
    print("\n=== Dynamic Class Creation Example ===")

    # Define methods that will be part of the dynamically created class
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def calculate(self):
        return self.x * self.y

    def __str__(self):
        return f"DynamicClass({self.x}, {self.y})"

    # Create a new class dynamically
    attributes = {
        "__init__": __init__,
        "calculate": calculate,
        "__str__": __str__,
    }

    DynamicClass = type("DynamicClass", (), attributes)

    # Create an instance of the dynamic class
    instance = DynamicClass(10, 20)
    print(f"Instance: {instance}")
    print(f"Calculation result: {instance.calculate()}")

    # Dynamically add a new method to the class
    def new_method(self):
        return self.x + self.y

    DynamicClass.add = new_method

    # Use the newly added method
    print(f"Addition result: {instance.add()}")


def main():
    print("Python Metaprogramming Demonstration")

    demonstrate_metaclass()
    demonstrate_class_decorator()
    demonstrate_singleton()
    demonstrate_dynamic_attributes()
    demonstrate_dynamic_class_creation()


if __name__ == "__main__":
    main()
