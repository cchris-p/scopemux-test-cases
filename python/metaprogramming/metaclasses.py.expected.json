{
  "language": "Python",
  "ast": {
    "type": "module",
    "functions": [
      {
        "name": "log_calls",
        "type": "function_definition",
        "parameters": [
          {"name": "cls", "type": "Any"}
        ],
        "return_type": "Any",
        "body_range": {
          "start_line": 16,
          "start_column": 0,
          "end_line": 35,
          "end_column": 4
        },
        "docstring": "A class decorator that logs method calls"
      },
      {
        "name": "singleton",
        "type": "function_definition",
        "parameters": [
          {"name": "cls", "type": "Any"}
        ],
        "return_type": "Any",
        "body_range": {
          "start_line": 38,
          "start_column": 0,
          "end_line": 48,
          "end_column": 4
        },
        "docstring": "A class decorator that implements the Singleton pattern"
      },
      {
        "name": "timing_method",
        "type": "function_definition",
        "parameters": [
          {"name": "func", "type": "Callable"}
        ],
        "return_type": "Callable",
        "body_range": {
          "start_line": 129,
          "start_column": 0,
          "end_line": 139,
          "end_column": 4
        },
        "docstring": "A method decorator that measures execution time"
      },
      {
        "name": "demonstrate_metaclass",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 242,
          "start_column": 0,
          "end_line": 266,
          "end_column": 4
        }
      },
      {
        "name": "demonstrate_class_decorator",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 269,
          "start_column": 0,
          "end_line": 280,
          "end_column": 4
        }
      },
      {
        "name": "demonstrate_singleton",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 283,
          "start_column": 0,
          "end_line": 296,
          "end_column": 4
        }
      },
      {
        "name": "demonstrate_dynamic_attributes",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 299,
          "start_column": 0,
          "end_line": 321,
          "end_column": 4
        }
      },
      {
        "name": "demonstrate_dynamic_class_creation",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 324,
          "start_column": 0,
          "end_line": 360,
          "end_column": 4
        }
      },
      {
        "name": "main",
        "type": "function_definition",
        "parameters": [],
        "return_type": "None",
        "body_range": {
          "start_line": 363,
          "start_column": 0,
          "end_line": 371,
          "end_column": 4
        }
      }
    ],
    "classes": [
      {
        "name": "ValidateAttributes",
        "type": "metaclass",
        "methods": [
          {
            "name": "__new__",
            "parameters": [
              {"name": "mcs", "type": "Type"},
              {"name": "name", "type": "str"},
              {"name": "bases", "type": "tuple"},
              {"name": "attrs", "type": "dict"}
            ],
            "return_type": "Type",
            "body_range": {
              "start_line": 53,
              "end_line": 64
            }
          },
          {
            "name": "__init__",
            "parameters": [
              {"name": "cls", "type": "Type"},
              {"name": "name", "type": "str"},
              {"name": "bases", "type": "tuple"},
              {"name": "attrs", "type": "dict"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 66,
              "end_line": 82
            }
          }
        ],
        "range": {
          "start_line": 51,
          "start_column": 0,
          "end_line": 82,
          "end_column": 8
        },
        "docstring": "A metaclass that enforces type validation for attributes"
      },
      {
        "name": "LazyProperty",
        "type": "descriptor",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "LazyProperty"},
              {"name": "func", "type": "Callable"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 87,
              "end_line": 90
            }
          },
          {
            "name": "__get__",
            "parameters": [
              {"name": "self", "type": "LazyProperty"},
              {"name": "instance", "type": "Any"},
              {"name": "owner", "type": "Type"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 92,
              "end_line": 102
            }
          }
        ],
        "range": {
          "start_line": 85,
          "start_column": 0,
          "end_line": 102,
          "end_column": 8
        },
        "docstring": "A descriptor that computes the value only when first accessed"
      },
      {
        "name": "TimingDescriptor",
        "type": "descriptor",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "TimingDescriptor"},
              {"name": "func", "type": "Callable"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 107,
              "end_line": 109
            }
          },
          {
            "name": "__get__",
            "parameters": [
              {"name": "self", "type": "TimingDescriptor"},
              {"name": "instance", "type": "Any"},
              {"name": "owner", "type": "Type"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 111,
              "end_line": 123
            }
          }
        ],
        "range": {
          "start_line": 105,
          "start_column": 0,
          "end_line": 123,
          "end_column": 8
        },
        "docstring": "A descriptor that measures the execution time of methods"
      },
      {
        "name": "AttributeProxy",
        "type": "descriptor",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "AttributeProxy"},
              {"name": "target_attr", "type": "str"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 144,
              "end_line": 145
            }
          },
          {
            "name": "__get__",
            "parameters": [
              {"name": "self", "type": "AttributeProxy"},
              {"name": "instance", "type": "Any"},
              {"name": "owner", "type": "Type"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 147,
              "end_line": 154
            }
          }
        ],
        "range": {
          "start_line": 142,
          "start_column": 0,
          "end_line": 154,
          "end_column": 8
        },
        "docstring": "A descriptor that delegates attribute access to another object"
      },
      {
        "name": "Person",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "Person"},
              {"name": "name", "type": "str"},
              {"name": "age", "type": "int"},
              {"name": "email", "type": "str"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 160,
              "end_line": 163
            }
          },
          {
            "name": "validate_name",
            "parameters": [
              {"name": "self", "type": "Person"},
              {"name": "value", "type": "Any"}
            ],
            "return_type": "str",
            "body_range": {
              "start_line": 165,
              "end_line": 169
            },
            "docstring": "Validate that name is a non-empty string"
          },
          {
            "name": "validate_age",
            "parameters": [
              {"name": "self", "type": "Person"},
              {"name": "value", "type": "Any"}
            ],
            "return_type": "int",
            "body_range": {
              "start_line": 171,
              "end_line": 175
            },
            "docstring": "Validate that age is a positive integer"
          },
          {
            "name": "validate_email",
            "parameters": [
              {"name": "self", "type": "Person"},
              {"name": "value", "type": "Any"}
            ],
            "return_type": "str",
            "body_range": {
              "start_line": 177,
              "end_line": 181
            },
            "docstring": "Validate that email contains @"
          },
          {
            "name": "formatted_info",
            "parameters": [
              {"name": "self", "type": "Person"}
            ],
            "return_type": "str",
            "body_range": {
              "start_line": 184,
              "end_line": 186
            },
            "docstring": "Return formatted information about the person",
            "decorator": "LazyProperty"
          },
          {
            "name": "do_work",
            "parameters": [
              {"name": "self", "type": "Person"},
              {"name": "duration", "type": "float", "default": "0.1"}
            ],
            "return_type": "str",
            "body_range": {
              "start_line": 189,
              "end_line": 191
            },
            "docstring": "Simulate doing some work",
            "decorator": "timing_method"
          }
        ],
        "metaclass": "ValidateAttributes",
        "range": {
          "start_line": 158,
          "start_column": 0,
          "end_line": 191,
          "end_column": 8
        },
        "docstring": "A person with validated attributes"
      },
      {
        "name": "Calculator",
        "type": "class_definition",
        "methods": [
          {
            "name": "add",
            "parameters": [
              {"name": "self", "type": "Calculator"},
              {"name": "a", "type": "Any"},
              {"name": "b", "type": "Any"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 198,
              "end_line": 199
            }
          },
          {
            "name": "subtract",
            "parameters": [
              {"name": "self", "type": "Calculator"},
              {"name": "a", "type": "Any"},
              {"name": "b", "type": "Any"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 201,
              "end_line": 202
            }
          },
          {
            "name": "multiply",
            "parameters": [
              {"name": "self", "type": "Calculator"},
              {"name": "a", "type": "Any"},
              {"name": "b", "type": "Any"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 204,
              "end_line": 205
            }
          },
          {
            "name": "divide",
            "parameters": [
              {"name": "self", "type": "Calculator"},
              {"name": "a", "type": "Any"},
              {"name": "b", "type": "Any"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 207,
              "end_line": 210
            }
          }
        ],
        "decorator": "log_calls",
        "range": {
          "start_line": 196,
          "start_column": 0,
          "end_line": 210,
          "end_column": 8
        },
        "docstring": "A simple calculator with logging"
      },
      {
        "name": "AppConfig",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "AppConfig"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 218,
              "end_line": 220
            }
          },
          {
            "name": "get_setting",
            "parameters": [
              {"name": "self", "type": "AppConfig"},
              {"name": "key", "type": "str"},
              {"name": "default", "type": "Any", "default": "None"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 222,
              "end_line": 223
            }
          },
          {
            "name": "set_setting",
            "parameters": [
              {"name": "self", "type": "AppConfig"},
              {"name": "key", "type": "str"},
              {"name": "value", "type": "Any"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 225,
              "end_line": 226
            }
          }
        ],
        "decorator": "singleton",
        "range": {
          "start_line": 215,
          "start_column": 0,
          "end_line": 226,
          "end_column": 8
        },
        "docstring": "A singleton configuration class"
      },
      {
        "name": "DynamicAttributes",
        "type": "class_definition",
        "methods": [
          {
            "name": "__init__",
            "parameters": [
              {"name": "self", "type": "DynamicAttributes"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 232,
              "end_line": 233
            }
          },
          {
            "name": "__getattr__",
            "parameters": [
              {"name": "self", "type": "DynamicAttributes"},
              {"name": "name", "type": "str"}
            ],
            "return_type": "Any",
            "body_range": {
              "start_line": 235,
              "end_line": 245
            },
            "docstring": "Called when an attribute is not found"
          },
          {
            "name": "__setattr__",
            "parameters": [
              {"name": "self", "type": "DynamicAttributes"},
              {"name": "name", "type": "str"},
              {"name": "value", "type": "Any"}
            ],
            "return_type": "None",
            "body_range": {
              "start_line": 247,
              "end_line": 253
            },
            "docstring": "Called when an attribute is set"
          }
        ],
        "range": {
          "start_line": 229,
          "start_column": 0,
          "end_line": 253,
          "end_column": 8
        },
        "docstring": "A class that handles missing attributes dynamically"
      }
    ],
    "imports": [
      {
        "name": "typing",
        "symbols": ["Any", "Dict", "List", "Callable", "Type", "Optional", "TypeVar", "cast"],
        "line": 11
      },
      {
        "name": "functools",
        "line": 12
      },
      {
        "name": "inspect",
        "line": 13
      },
      {
        "name": "time",
        "line": 14
      }
    ],
    "docstrings": [
      {
        "type": "module_docstring",
        "text": "Demonstration of Python Metaprogramming\n\nThis example covers:\n- Metaclasses\n- Class decorators\n- Dynamic attribute access (__getattr__, __setattr__)\n- Descriptors\n- Dynamic class creation (type)\n- Attribute access customization\n- Runtime class modification",
        "range": {
          "start_line": 1,
          "start_column": 0,
          "end_line": 10,
          "end_column": 3
        }
      }
    ],
    "decorators": [
      {
        "name": "LazyProperty",
        "applied_to": "Person.formatted_info",
        "line": 183
      },
      {
        "name": "timing_method",
        "applied_to": "Person.do_work",
        "line": 188
      },
      {
        "name": "log_calls",
        "applied_to": "Calculator",
        "line": 195
      },
      {
        "name": "singleton",
        "applied_to": "AppConfig",
        "line": 214
      }
    ],
    "metaprogramming_features": [
      {
        "feature": "metaclass",
        "examples": [
          {
            "class": "ValidateAttributes",
            "line": 51,
            "description": "Metaclass that enforces attribute validation"
          },
          {
            "class": "Person",
            "line": 158,
            "description": "Class using ValidateAttributes metaclass"
          }
        ]
      },
      {
        "feature": "descriptor",
        "examples": [
          {
            "class": "LazyProperty",
            "line": 85,
            "description": "Descriptor for lazy-loaded properties"
          },
          {
            "class": "TimingDescriptor",
            "line": 105,
            "description": "Descriptor for timing method execution"
          },
          {
            "class": "AttributeProxy",
            "line": 142,
            "description": "Descriptor for attribute delegation"
          }
        ]
      },
      {
        "feature": "class_decorator",
        "examples": [
          {
            "function": "log_calls",
            "line": 16,
            "description": "Decorator that logs method calls"
          },
          {
            "function": "singleton",
            "line": 38,
            "description": "Decorator that implements the Singleton pattern"
          }
        ]
      },
      {
        "feature": "dynamic_attribute_access",
        "examples": [
          {
            "method": "DynamicAttributes.__getattr__",
            "line": 235,
            "description": "Custom attribute lookup"
          },
          {
            "method": "DynamicAttributes.__setattr__",
            "line": 247,
            "description": "Custom attribute assignment"
          },
          {
            "method": "ValidateAttributes.__init__",
            "line": 66,
            "description": "Custom __setattr__ injection"
          }
        ]
      },
      {
        "feature": "dynamic_class_creation",
        "examples": [
          {
            "line": 351,
            "code": "DynamicClass = type('DynamicClass', (), attributes)",
            "description": "Creating a class dynamically using type()"
          },
          {
            "line": 357,
            "code": "DynamicClass.add = new_method",
            "description": "Adding a method to a class at runtime"
          }
        ]
      },
      {
        "feature": "method_wrapper",
        "examples": [
          {
            "line": 25,
            "code": "@functools.wraps(attr)",
            "description": "Method wrapping in log_calls"
          },
          {
            "line": 116,
            "code": "@functools.wraps(self.func)",
            "description": "Method wrapping in TimingDescriptor.__get__"
          }
        ]
      }
    ],
    "function_calls": [
      {
        "name": "main",
        "arguments": [],
        "line": 374
      }
    ],
    "dynamic_classes": [
      {
        "name": "DynamicClass",
        "created_at": 351,
        "methods_defined": [
          {"name": "__init__", "line": 328},
          {"name": "calculate", "line": 331},
          {"name": "__str__", "line": 334},
          {"name": "add", "line": 356}
        ],
        "instantiated_at": 354
      }
    ]
  }
}
